sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/ItemsScanBase.scala

```diff
@@ -0,0 +1,256 @@
+    override def readSchema(): StructType = {
+        schema
+    }
+
+    //scalastyle:off method.length
+    override def planInputPartitions(): Array[InputPartition] = {
+        if (plannedInputPartitionsRef.get() == null) {
+            val partitionMetadata = CosmosPartitionPlanner.getFilteredPartitionMetadata(
+                config,
+                clientConfiguration,
+                Some(cosmosClientStateHandles),
+                containerConfig,
+                partitioningConfig,
+                false
+            )
+
+            val calledFrom = s"ItemsScan($description()).planInputPartitions"
+            Loan(
+                List[Option[CosmosClientCacheItem]](
+                    Some(CosmosClientCache.apply(
+                        clientConfiguration,
+                        Some(cosmosClientStateHandles.value.cosmosClientMetadataCaches),
+                        calledFrom
+                    )),
+                    ThroughputControlHelper.getThroughputControlClientCacheItem(
+                        config, calledFrom, Some(cosmosClientStateHandles), sparkEnvironmentInfo)
+                ))
+                .to(clientCacheItems => {
+                    val container =
+                        ThroughputControlHelper.getContainer(
+                            config,
+                            containerConfig,
+                            clientCacheItems(0).get,
+                            clientCacheItems(1))
+                    SparkUtils.safeOpenConnectionInitCaches(container, log)
+
+                    val cosmosInputPartitions = CosmosPartitionPlanner
+                        .createInputPartitions(
+                            partitioningConfig,
+                            container,
+                            partitionMetadata,
+                            defaultMinPartitionCount,
+                            CosmosPartitionPlanner.DefaultPartitionSizeInMB,
+                            ReadLimit.allAvailable(),
+                            false
+                        )
+                    plannedInputPartitionsRef.set(cosmosInputPartitions)
+                })
+        }
+
+        if (readManyFilterMapRef.get() == null) {
+            // there is nothing to prune, return the original planned input partition
+            val afterPrunePlannedPartitions = plannedInputPartitionsRef.get().map(_.asInstanceOf[InputPartition])
+            if (afterPrunePlannedPartitions.size < plannedInputPartitionsRef.get().size) {
+                log.logInfo(s"There are ${plannedInputPartitionsRef.get().size - afterPrunePlannedPartitions.size} partitions got pruned")
+            }
+            afterPrunePlannedPartitions
+        } else {
+            // only return partitions has matching filter criteria
+            prunePartitions().map(_.asInstanceOf[InputPartition])
+        }
+    }
+
+    private def prunePartitions(): Array[CosmosInputPartition] = {
+        plannedInputPartitionsRef.get().filter(
```

Comment:
In theory it would be sufficient if input partition epk is a full subset of one of the readMany filters, correct?

Reply:
yea, here I check they match exactly because the ReadManyFilterAnalyzer will return a map which is the input partition feed range -> a list of filters, so using ==0 should be good here?

Reply:
updated